pkg/datastore/credentials.go

```diff
@@ -0,0 +1,73 @@
+package datastore
+
+import (
+	"context"
+	"fmt"
+	"sort"
+	"strings"
+
+	"github.com/aws/aws-sdk-go-v2/aws"
+	awsconfig "github.com/aws/aws-sdk-go-v2/config"
+	rdsauth "github.com/aws/aws-sdk-go-v2/feature/rds/auth"
+	"golang.org/x/exp/maps"
+
+	log "github.com/authzed/spicedb/internal/logging"
+)
+
+type CredentialsProvider interface {
+	// Get return the username and password to use when connecting to the database
+	Get(ctx context.Context, dbHostname string, dbPort uint16, dbUser string) (string, string, error)
+}
+
+type credentialsProviderBuilderFunc func(ctx context.Context) (CredentialsProvider, error)
+
+const (
+	AWSIAMCredentialProvider = "aws-iam"
+)
+
+var BuilderForCredentialProvider = map[string]credentialsProviderBuilderFunc{
+	AWSIAMCredentialProvider: newAWSIAMCredentialsProvider,
+}
+
+// CredentialsProviderOptions returns the full set of credential provider names, sorted and quoted into a string.
+func CredentialsProviderOptions() string {
+	ids := maps.Keys(BuilderForCredentialProvider)
+	sort.Strings(ids)
+	quoted := make([]string, 0, len(ids))
+	for _, id := range ids {
+		quoted = append(quoted, `"`+id+`"`)
+	}
+	return strings.Join(quoted, ", ")
+}
+
+// NewCredentialsProvider create a new CredentialsProvider for the given name
+// return nil if no match is found
+// return an error if there is a problem initializing the given CredentialsProvider
+func NewCredentialsProvider(ctx context.Context, name string) (CredentialsProvider, error) {
+	builder, ok := BuilderForCredentialProvider[name]
+	if !ok {
+		return nil, nil
+	}
+	return builder(ctx)
+}
+
+// AWS IAM provider
+
+func newAWSIAMCredentialsProvider(ctx context.Context) (CredentialsProvider, error) {
+	awsSdkConfig, err := awsconfig.LoadDefaultConfig(ctx)
+	if err != nil {
+		return nil, err
+	}
+	return &awsIamCredentialsProvider{awsSdkConfig: awsSdkConfig}, nil
+}
+
+type awsIamCredentialsProvider struct {
+	awsSdkConfig aws.Config
+}
+
+func (d awsIamCredentialsProvider) Get(ctx context.Context, dbHostname string, dbPort uint16, dbUser string) (string, string, error) {
+	dbEndpoint := fmt.Sprintf("%s:%d", dbHostname, dbPort)
+	authToken, err := rdsauth.BuildAuthToken(ctx, dbEndpoint, d.awsSdkConfig.Region, dbUser, d.awsSdkConfig.Credentials)
+	log.Ctx(ctx).Trace().Str("region", d.awsSdkConfig.Region).Str("endpoint", dbEndpoint).Str("user", dbUser).Msg("successfully retrieved IAM auth token for DB")
```

Comment:
return the error here if non-nil so we don't log a success message on failure

Reply:
good catch - corrected